### PR TITLE
feat: slashing

### PR DESCRIPTION
**Pull request by 0xShaito**
# 🤖 Linear

Closes GRT-XXX

### CODE REVIEW

File: src/contracts/HorizonAccountingExtension.sol [hunk 1]
```diff
@@ -175,17 +185,20 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+    _unbond(_pledger, _totalPledged);
 
-    /////// TODO: In here we have to pay the user the amount
-    // We can send it to their wallet or provision it again
+    // Send the user the amount they won by participating in the dispute
+    GRT.transfer(_pledger, _claimAmount - _totalPledged);
 
-    // pledgerClaimed[_requestId][_pledger] = true;
-    // balanceOf[_pledger] += _claimAmount;
+    pledgerClaimed[_requestId][_pledger] = true;
```
Comment (ashitakah) [inline src/contracts/HorizonAccountingExtension.sol:196]: As you said if we create a mapping with the requestId and the pledger we dont need to accounting the claimAmount

File: src/contracts/HorizonAccountingExtension.sol [hunk 2]
```diff
@@ -305,6 +318,56 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
     // emit Released(_requestId, _bonder, _token, _amount);
   }
 
+  function slash(bytes32 _disputeId, uint256 _nUsers) external {
```
Comment (ashitakah) [inline src/contracts/HorizonAccountingExtension.sol:321]: should has any access control?

File: src/contracts/HorizonAccountingExtension.sol [hunk 3]
```diff
@@ -305,6 +318,56 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+    if (_status != IOracle.DisputeStatus.NoResolution) {
+      _numberOfPledges = !(_status == IOracle.DisputeStatus.Won)
+        ? _result.bondEscalationModule.pledgesForDispute(_requestId, _pledger)
+        : _result.bondEscalationModule.pledgesAgainstDispute(_requestId, _pledger);
+    }
+
+    _slashAmount = _result.amountPerPledger * _numberOfPledges;
+  }
+
+  function _slash(bytes32 _disputeId, uint256 _nUsers) internal {
```
Comment (ashitakah) [inline src/contracts/HorizonAccountingExtension.sol:343]: I dont like nUsers as param

File: src/contracts/HorizonAccountingExtension.sol [hunk 4]
```diff
@@ -305,6 +318,56 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+
+      // Check if the user is actually slashable
+      uint256 _slashAmount = _calculateSlashAmount(_disputeId, _user);
+      if (_slashAmount > 0) {
+        // Slash the user
+        HORIZON_STAKING.slash(
+          _user,
+          _slashAmount,
+          _slashAmount, // TODO: How do we manage the max verifier cut?
+          // What if it's not 100%?
```
Comment (ashitakah) [inline src/contracts/HorizonAccountingExtension.sol:359]: wdym with this comment

File: src/contracts/HorizonAccountingExtension.sol [hunk 5]
```diff
@@ -305,6 +318,56 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+    _slashAmount = _result.amountPerPledger * _numberOfPledges;
+  }
+
+  function _slash(bytes32 _disputeId, uint256 _nUsers) internal {
+    EnumerableSet.AddressSet storage _users = _pledgers[_disputeId];
+
+    uint256 _slashedUsers;
+
+    while (_slashedUsers < _nUsers || _users.length() > 0) {
+      address _user = _users.at(0);
```
Comment (ashitakah) [inline src/contracts/HorizonAccountingExtension.sol:349]: is that right? You are always taking the same user

File: src/contracts/HorizonAccountingExtension.sol [hunk 6]
```diff
@@ -305,6 +318,56 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+        // Slash the user
+        HORIZON_STAKING.slash(
+          _user,
+          _slashAmount,
+          _slashAmount, // TODO: How do we manage the max verifier cut?
+          // What if it's not 100%?
+          address(this)
+        );
+
+        _slashedUsers++;
```
Comment (ashitakah) [inline src/contracts/HorizonAccountingExtension.sol:363]: unchecked maybe?

File: src/contracts/HorizonAccountingExtension.sol [hunk 7]
```diff
@@ -305,6 +318,56 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+  function _slash(bytes32 _disputeId, uint256 _nUsers) internal {
+    EnumerableSet.AddressSet storage _users = _pledgers[_disputeId];
+
+    uint256 _slashedUsers;
+
+    while (_slashedUsers < _nUsers || _users.length() > 0) {
+      address _user = _users.at(0);
+
+      // Check if the user is actually slashable
+      uint256 _slashAmount = _calculateSlashAmount(_disputeId, _user);
```
Comment (ashitakah) [inline src/contracts/HorizonAccountingExtension.sol:352]: should be named as pledger and not user?

File: src/contracts/HorizonAccountingExtension.sol [hunk 8]
```diff
@@ -305,6 +318,56 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+      _numberOfPledges = !(_status == IOracle.DisputeStatus.Won)
+        ? _result.bondEscalationModule.pledgesForDispute(_requestId, _pledger)
+        : _result.bondEscalationModule.pledgesAgainstDispute(_requestId, _pledger);
+    }
+
+    _slashAmount = _result.amountPerPledger * _numberOfPledges;
+  }
+
+  function _slash(bytes32 _disputeId, uint256 _nUsers) internal {
+    EnumerableSet.AddressSet storage _users = _pledgers[_disputeId];
```
Comment (ashitakah) [inline src/contracts/HorizonAccountingExtension.sol:344]: Maybe im missing something but, why we need to pass as param de nUsers to slash?

**Review by ashitakah — CHANGES_REQUESTED** [review]

File: src/contracts/HorizonAccountingExtension.sol (hunk 3 above)
Comment (0xShaito) [reply src/contracts/HorizonAccountingExtension.sol:343]: Any other ideas?

File: src/contracts/HorizonAccountingExtension.sol (hunk 5 above)
Comment (0xShaito) [reply src/contracts/HorizonAccountingExtension.sol:349]: But I'm removing the elemen from the set

File: src/contracts/HorizonAccountingExtension.sol (hunk 8 above)
Comment (0xShaito) [reply src/contracts/HorizonAccountingExtension.sol:344]: It's just to know how many users to slash

File: src/contracts/HorizonAccountingExtension.sol [hunk 9]
```diff
@@ -175,17 +185,20 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
     uint256 _claimAmount = _amountPerPledger * _numberOfPledges;
+    // TODO: We still need to try and slash some people in case there is not enough balance
+    // Unbond the pledged amount
+    uint256 _totalPledged = _result.bondSize * _numberOfPledges;
+    _unbond(_pledger, _totalPledged);
 
-    /////// TODO: In here we have to pay the user the amount
-    // We can send it to their wallet or provision it again
+    // Send the user the amount they won by participating in the dispute
+    GRT.transfer(_pledger, _claimAmount - _totalPledged);
```
Comment (0xJabberwock) [inline src/contracts/HorizonAccountingExtension.sol:194]: I suggest implementing `SafeERC20` for the token operations.

File: src/contracts/HorizonAccountingExtension.sol [hunk 10]
```diff
@@ -305,6 +318,56 @@ contract HorizonAccountingExtension is Validator, IHorizonAccountingExtension {
...
+
+    _slashAmount = _result.amountPerPledger * _numberOfPledges;
+  }
+
+  function _slash(bytes32 _disputeId, uint256 _nUsers) internal {
+    EnumerableSet.AddressSet storage _users = _pledgers[_disputeId];
+
+    uint256 _slashedUsers;
+
+    while (_slashedUsers < _nUsers || _users.length() > 0) {
```
Comment (0xJabberwock) [inline src/contracts/HorizonAccountingExtension.sol:348]: Shouldn't you use the logical AND (`&&`) for the `while` condition?

Comment (0xShaito) [reply src/contracts/HorizonAccountingExtension.sol:348]: GUT!

**Pull request closed without merging**